### ISSUE TITLE
Reporting Feature not working (Google Workspace)

### ISSUE THREAD

**Issue by Luke-From-IT** [gophish/gophish#2720]
**What version of Gophish are you using?**: 0.12.1 (Docker)

**Brief description of the issue**: IMAP reporting is not finding new emails in the Google Workspace email account.

**What are you expecting to see happen?**: IMAP reporting function is working, I have a phishing label set up in the google account and a filter to put reported phishing emails into this folder.

**What are you seeing happen?**
No new emails are found in the email account.

**Please provide any terminal output that may be relevant below (Email Redacted)**

```
time="2023-01-10T11:08:28Z" level=debug msg="Checking IMAP for user 2: dataleavers@XOSOS.co.uk -> imap.gmail.com"
time="2023-01-10T11:08:29Z" level=debug msg="No new emails for dataleavers@XOSOS.co.uk"
time="2023-01-10T11:09:29Z" level=debug msg="Checking IMAP for user 2: dataleavers@XOSOS.co.uk -> imap.gmail.com"
time="2023-01-10T11:09:29Z" level=debug msg="No new emails for dataleavers@XOSOS.co.uk"
time="2023-01-10T11:10:27Z" level=info msg="66.249.92.152 - - [10/Jan/2023:11:10:27 +0000] \"GET /track?rid=1rbuW39 HTTP/1.1\" 200 95 \"http://mail.google.com/\" \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.246 Mozilla/5.0\""
time="2023-01-10T11:10:29Z" level=debug msg="Checking IMAP for user 2: dataleavers@XOSOS.co.uk -> imap.gmail.com"
time="2023-01-10T11:10:29Z" level=debug msg="No new emails for dataleavers@XOSOS.co.uk"
time="2023-01-10T11:10:32Z" level=info msg="81.174.135.167 - - [10/Jan/2023:11:10:32 +0000] \"GET /track?rid=1rbuW39 HTTP/1.1\" 200 95 \"\" \"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko)\""
time="2023-01-10T11:10:35Z" level=info msg="81.174.135.167 - - [10/Jan/2023:11:10:35 +0000] \"GET /track?rid=1rbuW39 HTTP/1.1\" 200 95 \"\" \"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko)\""
time="2023-01-10T11:11:04Z" level=info msg="66.249.92.153 - - [10/Jan/2023:11:11:04 +0000] \"GET /track?rid=1rbuW39 HTTP/1.1\" 200 95 \"http://mail.google.com/\" \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.246 Mozilla/5.0\""
time="2023-01-10T11:11:08Z" level=info msg="66.249.92.152 - - [10/Jan/2023:11:11:08 +0000] \"GET /track?rid=1rbuW39 HTTP/1.1\" 200 95 \"http://mail.google.com/\" \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.246 Mozilla/5.0\""
time="2023-01-10T11:11:08Z" level=info msg="66.249.92.152 - - [10/Jan/2023:11:11:08 +0000] \"GET /track?rid=1rbuW39 HTTP/1.1\" 200 95 \"http://mail.google.com/\" \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.246 Mozilla/5.0\""
time="2023-01-10T11:11:09Z" level=info msg="66.249.92.151 - - [10/Jan/2023:11:11:09 +0000] \"GET /track?rid=1rbuW39 HTTP/1.1\" 200 95 \"http://mail.google.com/\" \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.246 Mozilla/5.0\""
time="2023-01-10T11:11:29Z" level=debug msg="Checking IMAP for user 2: dataleavers@XOSOS.co.uk -> imap.gmail.com"
time="2023-01-10T11:11:29Z" level=debug msg="No new emails for dataleavers@XOSOS.co.uk"
```

**Please provide as many steps as you can to reproduce the problem:**
- Set up IMAP reporting with gmail account
- Set up campaign
- Report email by forwarding to phishing@DOMAIN
- Confirm email has reached the gmail account and filtered into the gmail label
- Check campaign to see if report has been updated (shows 0)
- Check gophish logs to see if any new emails are reported (shows 0)
 
Reporting Settings in GoPhish (Email Redacted)
<img width="943" alt="image" src="https://user-images.githubusercontent.com/46346663/211535878-6f9d5324-251b-4b32-91fb-e4cb226f8b78.png">

Emails in Google Account under the label (note search says "phishing", but the actual label is "Phishing" and confirmed this by changing in gophish reporting section.
<img width="1430" alt="image" src="https://user-images.githubusercontent.com/46346663/211536429-532b04cf-88cf-41b5-9f0a-55600c2d7f29.png">

**Comment by glennzw** [comment]
Hi @Luke-From-IT (cool handle btw) - thanks for reaching out. Are the emails in the Phishing folder unread? In your screenshot it looks like they're already read, but maybe that's because you've manually had a look after gophish IMAP failed.

Also, if you skip the Phishing folder and just use the inbox in your gophish IMAP settings, does that work? I'm not sure how GMail distinguishes between folders/labels.

**Comment by Luke-From-IT** [comment]
@glennzw, thanks for the response.

I have a google filter set up to process the emails into that folder, they were marked as unread, but I also removed the action to mark them as unread as still the same result. I also tried directly from the inbox, but there were also no results.

I host gophish behind traefik reverse proxy (so it is on port 80 redirecting to port 443). Still, there are no additional middlewares on the gophish service (only the gophish admin UI service is behind Google OAuth).

I believe the below is a separate issue, but:
It also looks like the IMAP check has stopped recently, the last update being over an hour and a half ago, I also noticed when I update the polling frequency from 60 to another number (5), and save it resets immediately to 60 and still does not poll the IMAP folder any more.

p.s. the name was intended to stop me from getting tagged in issues for one of the developers at my company who has a similar name... still doesn't help though haha!

**Comment by glennzw** [comment]
Thanks for the extra info and patience with figuring it out. I'm testing this on my side and observe the following with GMail:

* Polling INBOX and new emails are observed (campaigns identified, non campaigns marked a such)
* Polling Phishing folder and unread emails are *not* observed

I'm not sure if this is a weird artefact of the way GMail handles folders (as it calls them 'Labels') but I'm doing some more digging now.

P.S Same this side, but we call the other Glenn 'Bob' now :)

**Comment by glennzw** [comment]
Hmm it seems that it's definitely checking the appropriate folder as trying a folder that doesn't exist gives us an error

> time="2023-01-13T14:46:23Z" level=error msg="failed to create IMAP connection: Unknown Mailbox: Derpfolder (Failure)"

**Comment by glennzw** [comment]
Just tested with Outlook.com, and checking an IMAP folder called 'Phishing' works:

<img width="778" alt="Screenshot 2023-01-14 at 11 10 08" src="https://user-images.githubusercontent.com/3966613/212469120-1079590e-8d81-4f93-82f0-2377efcc4a29.png">


```
time="2023-01-14T11:09:37Z" level=debug msg="Checking IMAP for user 26: gophish-test@outlook.com -> outlook.live.com"
time="2023-01-14T11:09:39Z" level=debug msg="2 new emails for gophish-test@outlook.com"
time="2023-01-14T11:09:39Z" level=info msg="User 'Safety Check <safety.check.9001@gmail.com>' reported email with rid fDW4e4n"
```

I'm going to try set up some MiTM plumbing to see how the Google conversation differs.

**Comment by glennzw** [comment]
tl; dr: Don't use the label 'Phishing' with GMail, it seems to be a reserved word. Keep reading for more info.

--
Just to keep you in the loop @Luke-From-IT I've been running some MiTM tests to try figure out why GMail is behaving like this.

First I setup a socat MiTM tunnel:

`socat -v tcp4-listen:993,reuseaddr,fork ssl:imap.gmail.com:993,verify=0`

### 1. GMail - Phishing folder
```
* OK Gimap ready for requests from 2.1.77.0 xxxxxxxx
SePssw CAPABILITY
* CAPABILITY IMAP4rev1 UNSELECT IDLE NAMESPACE QUOTA ID XLIST CHILDREN X-GM-EXT-1 XYZZY SASL-IR AUTH=XOAUTH2 AUTH=PLAIN AUTH=PLAIN-CLIENTTOKEN AUTH=OAUTHBEARER AUTH=XOAUTH
SePssw OK Thats all she wrote! xxxxxx
6MAkpA LOGIN "safety.check.9001@gmail.com" "xxxxxxxxxxxx"
* CAPABILITY IMAP4rev1 UNSELECT IDLE NAMESPACE QUOTA ID XLIST CHILDREN X-GM-EXT-1 UIDPLUS COMPRESS=DEFLATE ENABLE MOVE CONDSTORE ESEARCH UTF8=ACCEPT LIST-EXTENDED LIST-STATUS LITERAL- SPECIAL-USE APPENDLIMIT=35651584
6MAkpA OK safety.check.9001@gmail.com authenticated (Success)
47VW4g SELECT "Phishing"
* FLAGS (\\Answered \\Flagged \\Draft \\Deleted \\Seen $NotJunk $NotPhishing $Phishing NotJunk)
* OK [PERMANENTFLAGS (\\Answered \\Flagged \\Draft \\Deleted \\Seen $NotJunk $NotPhishing $Phishing NotJunk \\*)] Flags permitted.
* OK [UIDVALIDITY 14] UIDs valid.
* 2 EXISTS
* 0 RECENT
* OK [UIDNEXT 3] Predicted next UID.
* OK [HIGHESTMODSEQ 642752]
47VW4g OK [READ-WRITE] Phishing selected. (Success)
Zh5FWQ SEARCH CHARSET UTF-8 UNSEEN
* SEARCH
Zh5FWQ OK SEARCH completed (Success)
zpZMkA LOGOUT
* BYE LOGOUT Requested
zpZMkA OK 73 good day (Success)
```

So it's logging in, selecting the 'Phishing' folder, and noticing that there are two unread emails (`* 2 EXISTS`) but then not getting any emails back. Compared to MiTM'ing Outlook.com:

### 2. Outlook - Phishing folder

```
* OK The Microsoft Exchange IMAP4 service is ready. [xxxxxxxxx==]
nQviuQ CAPABILITY
* CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
nQviuQ OK CAPABILITY completed.
j9PwvA LOGIN "gophish-test@outlook.com" "xxxxxxxxxxx"
j9PwvA OK LOGIN completed.
Ob6UdA SELECT "Phishing"
* 2 EXISTS
* 0 RECENT
* FLAGS (\\Seen \\Answered \\Flagged \\Deleted \\Draft $MDNSent)
* OK [PERMANENTFLAGS (\\Seen \\Answered \\Flagged \\Deleted \\Draft $MDNSent)] Permanent flags
* OK [UNSEEN 1] Is the first unseen message
* OK [UIDVALIDITY 272] UIDVALIDITY value
* OK [UIDNEXT 3] The next unique identifier value
Ob6UdA OK [READ-WRITE] SELECT completed.
pvtgnw SEARCH CHARSET UTF-8 UNSEEN
pvtgnw NO [BADCHARSET (US-ASCII)] The specified charset is not supported.
z2hcVA SEARCH CHARSET US-ASCII UNSEEN
* SEARCH 1
z2hcVA OK SEARCH completed.
4XDC6Q FETCH 1 (ENVELOPE FLAGS INTERNALDATE BODY[])
* 1 FETCH (ENVELOPE ("Tue, 10 May 2022 21:53:29 +0000" "FW: Outlook size ....... etc emails follow
```

With Outlook above we get the same SEARCH and 2 unread email response, but the emails come with.

### 3. GMail - Inbox folder

```
* OK Gimap ready for requests from 1.1.1.1 xxx
VBVLMg CAPABILITY
* CAPABILITY IMAP4rev1 UNSELECT IDLE NAMESPACE QUOTA ID XLIST CHILDREN X-GM-EXT-1 XYZZY SASL-IR AUTH=XOAUTH2 AUTH=PLAIN AUTH=PLAIN-CLIENTTOKEN AUTH=OAUTHBEARER AUTH=XOAUTH
VBVLMg OK Thats all she wrote! b8mb31854156wrm
8Inl3w LOGIN "safety.check.9001@gmail.com" "xxxxx"
* CAPABILITY IMAP4rev1 UNSELECT IDLE NAMESPACE QUOTA ID XLIST CHILDREN X-GM-EXT-1 UIDPLUS COMPRESS=DEFLATE ENABLE MOVE CONDSTORE ESEARCH UTF8=ACCEPT LIST-EXTENDED LIST-STATUS LITERAL- SPECIAL-USE APPENDLIMIT=35651584
8Inl3w OK safety.check.9001@gmail.com authenticated (Success)
7slWUA SELECT Inbox
* FLAGS (\\Answered \\Flagged \\Draft \\Deleted \\Seen $NotJunk $NotPhishing $Phishing NotJunk)
* OK [PERMANENTFLAGS (\\Answered \\Flagged \\Draft \\Deleted \\Seen $NotJunk $NotPhishing $Phishing NotJunk \\*)] Flags permitted.
* OK [UIDVALIDITY 1] UIDs valid.
* 166 EXISTS
* 0 RECENT
* OK [UIDNEXT 201] Predicted next UID.
* OK [HIGHESTMODSEQ 642876]
7slWUA OK [READ-WRITE] Inbox selected. (Success)
sabDJA SEARCH CHARSET UTF-8 UNSEEN
* SEARCH 157
sabDJA OK SEARCH completed (Success)
xtsQRQ FETCH 157 (ENVELOPE FLAGS INTERNALDATE BODY[])
* 157 FETCH (INTERNALDATE "15-Feb-2022 08:56:29 +0000" FLAGS (\\Seen) ENVELOPE ("Tue, 15 Feb 2022 08:56:28 +0000" "FW: Updates to our terms of use
*
 ```

Using the inbox with GMail we get the unread emails picked up. One thing I noticed in the above output is:

`* FLAGS (\\Answered \\Flagged \\Draft \\Deleted \\Seen $NotJunk $NotPhishing $Phishing NotJunk)`

....which makes me think maybe 'Phishing' is a reserved folder/label name? Let's test that theory with a random folder name:


### 4.  - GMail - RandomFolderName folder

Aha!

```
* OK Gimap ready for requests from 1.1.1.1
aEq72w CAPABILITY
* CAPABILITY IMAP4rev1 UNSELECT IDLE NAMESPACE QUOTA ID XLIST CHILDREN X-GM-EXT-1 XYZZY SASL-IR AUTH=XOAUTH2 AUTH=PLAIN AUTH=PLAIN-CLIENTTOKEN AUTH=OAUTHBEARER AUTH=XOAUTH
aEq72w OK Thats all she wrote! xx
7rPsAA LOGIN "safety.check.9001@gmail.com" "xxx"
* CAPABILITY IMAP4rev1 UNSELECT IDLE NAMESPACE QUOTA ID XLIST CHILDREN X-GM-EXT-1 UIDPLUS COMPRESS=DEFLATE ENABLE MOVE CONDSTORE ESEARCH UTF8=ACCEPT LIST-EXTENDED LIST-STATUS LITERAL- SPECIAL-USE APPENDLIMIT=35651584
7rPsAA OK safety.check.9001@gmail.com authenticated (Success)
BIVlkQ SELECT "RandomFolderName"
* FLAGS (\\Answered \\Flagged \\Draft \\Deleted \\Seen $NotJunk $NotPhishing $Phishing NotJunk)
* OK [PERMANENTFLAGS (\\Answered \\Flagged \\Draft \\Deleted \\Seen $NotJunk $NotPhishing $Phishing NotJunk \\*)] Flags permitted.
* OK [UIDVALIDITY 15] UIDs valid.
* 2 EXISTS
* 0 RECENT
* OK [UIDNEXT 3] Predicted next UID.
* OK [HIGHESTMODSEQ 642988]
BIVlkQ OK [READ-WRITE] RandomFolderName selected. (Success)
3ljtFQ SEARCH CHARSET UTF-8 UNSEEN
* SEARCH 1 2
3ljtFQ OK SEARCH completed (Success)
8eL7Ew FETCH 1:2 (ENVELOPE FLAGS INTERNALDATE BODY[])
* 1 FETCH (INTERNALDATE "18-May-2020 09:00:27 +0000" FLAGS (\\Seen) ENVELOPE ("Mon, 18 May 2020 08:55:25 GMT" "Critical security alert"  
* 
```


.... so that worked! OK so it seems that if you name a folder 'Phishing' weird things will happen. I'll try figure out if this is a bug and report it, or a feature and we'll just have to not using the label/folder 'Phishing' for GMail.

Thanks for your patience, and let me know if this works for you!

**Comment by Luke-From-IT** [comment]
@glennzw great find. It seems Gmail has some additional flags there, so it's picking it up as a folder name?

I have an additional issue that seems related, where the IMAP settings aren't saving, and there is no logs to show that IMAP is working.

You can see from the attached video that the Polling frequency doesn't update, and nothing shows in the logs either.

https://user-images.githubusercontent.com/46346663/212687273-a522c4c2-43fc-4f47-aa52-781b288f0d71.mp4

**Comment by glennzw** [comment]
> @glennzw great find. It seems Gmail has some additional flags there, so it's picking it up as a folder name?

To be honest I'm not sure - but I guess for now avoid using the keyword 'Phishing'. I'll poke some friends who work at Google.

Hmm now as for the IMAP settings issue - my first thought is that I'd be a bit nervous to poll every 5 seconds. Googling it seems GMail may rate limit you:

> Make sure your mail client isn't set to check for new mail too often. If your mail client checks for new messages more than once every 10 minutes, your client might repeatedly request your username and password.

(https://support.google.com/mail/answer/7126229?visit_id=638094779436167939-3670165855&rd=2#cantsignin&zippy=%2Ci-cant-sign-in-to-my-email-client)

If it's a default gophish setup the settings will be stored in `gophish.db` which you can edit/view with `sqlite3`. e.g:

```bash
➜  gophish git:(master) ✗ sqlite3 gophish.db
Enter ".help" for usage hints.
sqlite> SELECT * FROM imap;
26|localhost|993|gophish-test@outlook.com|somepassword|2023-01-14 15:38:49.063845+00:00|0|0|Phishing||0|0001-01-01 00:00:00+00:00|60|0
1|localhost|993|safety.check.9001@gmail.com|somepassword|2023-01-16 15:00:12.689262+00:00|0|0|RandomFolderName||1|0001-01-01 00:00:00+00:00|60|1
sqlite> 
```

You're right on the `2022 years ago` nonsensical value - I think that snuck in when doing some other work. I'll investigate. As you can see in the db values above the last successful login is marked as `0001-01-01 00:00:00+00:00` for some reason.

**Comment by glennzw** [comment]
Actually I can't reproduce this. The one thing to keep in mind is that whenever you save the IMAP settings or Disable/Enable IMAP in the UI the 'Last Login' date gets set back to the default `0001-01-01 00:00:00+00:00`.

But as soon as I see a successful login, I see the tracker update:

![Screenshot 2023-01-16 at 15 39 05](https://user-images.githubusercontent.com/3966613/212716286-f1dfdd60-d1ef-49b8-9559-9dbe96cd1fd1.png)

and reflect in the database:

```
1|localhost|993|safety.check.9001@gmail.com|somepassword|2023-01-16 15:34:48.693723+00:00|0|1|Inbox||1|2023-01-16 15:35:13.640069+00:00|60|1
```

**Comment by Luke-From-IT** [comment]
> Hmm now as for the IMAP settings issue - my first thought is that I'd be a bit nervous to poll every 5 seconds. Googling it seems GMail may rate limit you:

This was more of an example to see if anything shows in the logs :), after running a DELETE on the IMAP database table and setting it up again, it is now working and polling correctly :)

`time="2023-01-16T17:00:46Z" level=debug msg="Checking IMAP for user 2: dataleavers@carwow.co.uk -> imap.gmail.com"
time="2023-01-16T17:00:47Z" level=debug msg="1 new emails for dataleavers@domain.co.uk"
time="2023-01-16T17:00:47Z" level=info msg="User 'Luke <xyz@domain.co.uk>' reported email with rid coa6Jgy"
time="2023-01-16T17:00:47Z" level=error msg="Error reporting GoPhish email with rid coa6Jgy: record not found"
time="2023-01-16T17:00:47Z" level=debug msg="Marking 1 email as unread as failed to report"`

I will leave the sqlite3 commands here for anyone else who stumbles upon this, firstly stop the go phish service (or docker container).

Open the sqlite3 Database.
`sqlite3 gophish.db`

Query the IMAP table
`SELECT * from imap;`

Delete the existing IMAP table contents
`DELETE FROM imap;`

Query the IMAP table again to make sure everything is removed
`SELECT * FROM imap;`

Exit sqlite3
`.quit`

**Comment by Luke-From-IT** [comment]
Happy to close this now 👍 thanks for the quick and intuitive support

**Comment by glennzw** [comment]
You're very welcome Luke from IT!
